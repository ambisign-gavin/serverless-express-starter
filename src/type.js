// @flow

export type Args = {|
    name: string,
|}

export type PackageManagerPlatform = 'npm' | 'yarn';

export type TypeCheckerPlatform = 'none' | 'flow' | 'typescript';