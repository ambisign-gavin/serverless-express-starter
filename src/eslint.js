// @flow
import execa from 'execa';
import fs from 'fs';
import { join } from 'path';

class Eslint {
    getPackages(): Array<string> {
        return [
            'eslint@^5.0.0',
            'babel-eslint@^10.0.0'
        ];
    }

    async init(projectPath: string): execa.ExecaChildProcess {
        return execa('node_modules/eslint/bin/eslint.js',[
            '--init'
        ], {
            cwd: projectPath,
            stdio: 'inherit'
        });
    }

    rejectParserConfig(projectPath: string) {
        const jsConfig = join(projectPath, '.eslintrc.js');
        if (fs.existsSync(jsConfig)) {
            let eslintrc: { default: Object, parser: string } = require(jsConfig);
            delete eslintrc['default'];
            eslintrc['parser'] = 'babel-eslint';
            const newEslintrc = 'module.exports = ' + JSON.stringify(eslintrc, null, 4) + '\n';
            fs.writeFileSync(jsConfig, newEslintrc);
        }

        const jsonConfig = join(projectPath, '.eslintrc.json');
        if (fs.existsSync(jsonConfig)) {
            let eslintrc = require(jsonConfig);
            eslintrc = {
                ...eslintrc,
                parser: 'babel-eslint'
            };
            const newEslintrc = JSON.stringify(eslintrc, null, 4) + '\n';
            fs.writeFileSync(jsonConfig, newEslintrc);
        }
    }
}

const eslint = new Eslint();

export default eslint;