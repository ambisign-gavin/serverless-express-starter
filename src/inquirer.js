// @flow
import inquirer, { type Questions } from 'inquirer';
import { type PackageManagerPlatform, type TypeCheckerPlatform } from './type';

async function inquer(questions: Questions) {
    const answers = await inquirer.prompt([questions]);
    return answers[questions.name];
}

export class InquirerRobot {
    _name: string = '';
    _description: string = '';
    _pkgManager: PackageManagerPlatform = 'npm';
    _typeChecker: TypeCheckerPlatform = 'none';
    _isUsedEslint: boolean = false;

    async run() {
        this._name = await inquer({
            name: 'ProjectName',
            message: 'What\'s name for this project?'
        });

        this._description = await inquer({
            name: 'ProjectDescription',
            message: 'The project description'
        });

        this._pkgManager = await inquer({
            type: 'list',
            name: 'PackageManager',
            message: 'npm or yarn',
            choices: [
                'npm',
                'yarn'
            ]
        });

        this._typeChecker = await inquer({
            type: 'list',
            name: 'TypeChecker',
            message: 'Do you want to use type checker?',
            choices: [
                'none',
                'flow'
            ]
        });

        this._isUsedEslint = await inquer({
            type: 'confirm',
            name: 'UseEslint',
            message: 'Do you want to use eslint?'
        });
    }

    get name(): string {
        return this._name;
    }

    get description(): string {
        return this._description;
    }

    get pkgManager(): PackageManagerPlatform {
        return this._pkgManager;
    }

    get typeChecker(): TypeCheckerPlatform {
        return this._typeChecker;
    }

    get isUsedEslint(): boolean {
        return this._isUsedEslint;
    }
}

const inquirerRobot = new InquirerRobot();

export default inquirerRobot;
