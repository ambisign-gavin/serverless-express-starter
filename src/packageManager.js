// @flow
import execa from 'execa';
import { type PackageManagerType } from './type';

const dependenciesModule = [
    'express@^4.0.0',
    'serverless-http@^1.0.0',
];

const devDependenciesModule = [
    '@babel/cli@^7.0.0',
    '@babel/core@^7.0.0',
    '@babel/plugin-transform-modules-commonjs@^7.0.0',
    'cli-confirm@^1.0.0',
    'serverless-offline@^4.0.0',
];

class PackageManager {
    async install(packageManager: PackageManagerType, projectPath: string): Promise<void> {
        try {
            await execa(packageManager,[
                'add',
                ...dependenciesModule
            ], {
                cwd: projectPath,
                stdio: 'inherit'
            });
        
            await execa(packageManager,[
                'add',
                '-D',
                ...devDependenciesModule
            ], {
                cwd: projectPath,
                stdio: 'inherit'
            });
        } catch (error) {
            return Promise.reject(error);
        }
        return Promise.resolve();
    }

    generatePackageJson(packageManager: PackageManagerType, name: string, description: string): Object {
        const packageJson = {
            name,
            main: 'index.js',
            version: '1.0.0',
            description,
            scripts: {
                'build': 'babel src/ -d lib/',
                'local': `${packageManager} run build && sls offline start`,
                'deploy:dev': `${packageManager} run build && serverless deploy --stage dev`,
                'confirm:production': 'cli-confirm \' Do you want to deploy production server? \'',
                'deploy:production': `${packageManager} run confirm:production && ${packageManager} run build && serverless deploy --stage production`
            },
            dependencies: {},
            devDependencies: {},
            keywords: [],
        };
        return packageJson;
    }

}

const packageInstaller = new PackageManager();

export default packageInstaller;