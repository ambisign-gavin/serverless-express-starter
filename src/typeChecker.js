// @flow
import { type TypeCheckerPlatform } from './type';
import execa from 'execa';

class TypeChecker {
    generateTemplatePath(typeCheckerPlatform: TypeCheckerPlatform): string {
        switch (typeCheckerPlatform) {
        case 'none':
            return 'template/default/';
        case 'flow':
            return 'template/flow/';
        default:
            return 'template/default/';
        }
    }

    getInstallPackage(typeCheckerPlatform: TypeCheckerPlatform): Array<string> {
        if (typeCheckerPlatform === 'flow') {
            return [
                'flow-bin@^0.95.0',
                'flow-typed@^2.0.0',
                '@babel/preset-flow'
            ];
        }
        return [];
    }
    
    async runExtraSettings(typeCheckerPlatform: TypeCheckerPlatform, projectPath: string) {
        if (typeCheckerPlatform === 'flow') {
            await execa('flow-typed', [
                'install',
                'express@^4.0.0'
            ], {
                cwd: projectPath
            });
        }
    }
}

const typeChecker = new TypeChecker();

export default typeChecker;