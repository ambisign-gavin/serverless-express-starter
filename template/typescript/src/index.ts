import serverless from 'serverless-http';
import helloController from './hello';

export const hello = serverless(helloController());