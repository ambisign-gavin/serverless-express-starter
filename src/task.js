// @flow
import chalk from 'chalk';
import { writeFileSync } from 'fs';
import del from 'del';
import { join } from 'path';
import execa from 'execa';
import { InquirerRobot } from './tool/inquirer';
import packageManager from './tool/packageManager';
import eslint from './tool/eslint';

class Task {
    async createFiles(inquirerRobot: InquirerRobot) {
        try {
            const projectName = inquirerRobot.name;
            const projectPath = join(process.cwd(), inquirerRobot.name);
            const pkgManagerPlatform = inquirerRobot.pkgManager;
            const typeChecker = inquirerRobot.typeChecker;
    
            await execa('git', [
                'clone',
                'https://github.com/ambisign-gavin/serverless-express-starter.git',
                projectName
            ]);
    
            await del([
                join(projectPath, 'src'),
                join(projectPath, 'bin'),
                join(projectPath, '.gitignore'),
                join(projectPath, '.eslintrc.js'),
                join(projectPath, '.flowconfig'),
                join(projectPath, '.babelrc'),
                join(projectPath, '.git'),
                join(projectPath, 'package-lock.json'),
            ]);
    
            execa.shellSync(`cp -a ${typeChecker.generateTemplatePath()}. ./`, {
                cwd: projectPath
            });
    
            await del([
                join(projectPath, 'template'),
            ]);
    
            const packageContext = packageManager.generatePackageJson(
                pkgManagerPlatform,
                projectName,
                inquirerRobot.description
            );
    
            writeFileSync(join(projectPath, 'package.json'), JSON.stringify(packageContext, null, 2) + '\n');
        } catch (error) {
            console.log('An error occurred while creating files:', error);
        }
        
    }
    
    async installPackages(inquirerRobot: InquirerRobot) {
        try {
            let dependenciesModule = [
                'express@^4.0.0',
                'serverless-http@^1.0.0',
            ];
            
            let devDependenciesModule = [
                '@babel/cli@^7.0.0',
                '@babel/core@^7.0.0',
                '@babel/plugin-transform-modules-commonjs@^7.0.0',
                'cli-confirm@^1.0.0',
                'serverless-offline@^4.0.0',
            ];
    
            const projectPath = join(process.cwd(), inquirerRobot.name);
            const pkgManagerPlatform = inquirerRobot.pkgManager;
            const typeChecker = inquirerRobot.typeChecker;
    
            if (typeChecker.getInstallPackages().length > 0) {
                devDependenciesModule.push(...typeChecker.getInstallPackages());
            }
            
            if (inquirerRobot.isUsedEslint) {
                devDependenciesModule.push(...eslint.getInstallPackages());
            }
    
            await packageManager.install(
                pkgManagerPlatform,
                projectPath,
                dependenciesModule,
                false,
            );
    
            await packageManager.install(
                pkgManagerPlatform,
                projectPath,
                devDependenciesModule,
                true,
            );
        } catch (error) {
            console.log('An error occurred while installing packages:', error);
        }
        
    }
    
    async runExtraSettings(inquirerRobot: InquirerRobot) {
        try {
            const projectPath = join(process.cwd(), inquirerRobot.name);
            const typeChecker = inquirerRobot.typeChecker;

            await typeChecker.runExtraSettings(projectPath);

            if (inquirerRobot.isUsedEslint) {
                await eslint.init(projectPath);
                eslint.injectionParserConfig(projectPath);
                await eslint.fixScripts(projectPath);
            }
        } catch (error) {
            console.log('An error occurred while setting plugins:', error);
        }
        
    }
    
    showCompleteMessages() {
        const messages = `
It's all done! You can start to develop your mighty project!

You can use:

  npm/yarn run local               Run the local server.
  npm/yarn run deploy:dev          Deploy the server of dev stage.
  npm/yarn run deploy:production   Deploy the server of production stage.

Enjoy it!
        `;
        console.log(chalk.green(messages));
    }
}

const task = new Task();

export default task;