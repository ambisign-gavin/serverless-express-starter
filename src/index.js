// @flow
import { argv } from 'yargs';
import { commandChecker, initProject } from './task';

export function run() {
    if (!commandChecker(argv)) {
        return;
    }
    initProject(argv);
}


