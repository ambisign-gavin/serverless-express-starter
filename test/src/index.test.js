// @flow
import yargs from 'yargs';
import { run } from '../../src';
import execa from 'execa';
import task from '../../src/task';
import inquirerRobot from '../../src/tool/inquirer';

jest.mock('yargs');
jest.mock('execa');
jest.mock('../../src/task');
jest.mock('chalk', () => ({
    red: jest.fn().mockImplementation(text => text),
}));

function spyInquirerRobotDefaultValue() {
    jest.spyOn(inquirerRobot, 'name', 'get').mockReturnValue('my-server');
    jest.spyOn(inquirerRobot, 'pkgManager', 'get').mockReturnValue('npm');
    jest.spyOn(inquirerRobot, 'typeChecker', 'get').mockReturnValue('none');
    jest.spyOn(inquirerRobot, 'isUsedEslint', 'get').mockReturnValue(false);
    jest.spyOn(inquirerRobot, 'run').mockReturnValue(Promise.resolve());
}

describe('Runing command with options', () => {

    beforeAll(() => {
        jest.clearAllMocks();
        jest.spyOn(Object, 'keys').mockReturnValueOnce([
            '1',
            '2',
            '3'
        ]);
    });

    it('should show help when unknow options', async () => {
        
        let log = jest.spyOn(console, 'log');
        await run();
        expect(log.mock.calls.length).toEqual(2);
        expect(log.mock.calls[0][0]).toMatchSnapshot();
        expect(log.mock.calls[1][0]).toMatchSnapshot();
        expect(yargs.showHelp.mock.calls.length).toEqual(1);
    });

});

describe('Runing command without git', () => {

    beforeAll(() => {
        jest.clearAllMocks();
        jest.spyOn(Object, 'keys').mockReturnValueOnce([]);
        execa.shellSync.mockImplementation(() => {
            throw 'Error';
        });
    });

    it('should show error when not installed git', async () => {
        let log = jest.spyOn(console, 'log');
        await run();
        expect(execa.shellSync.mock.calls.length).toEqual(1);
        expect(log.mock.calls.length).toEqual(1);
        expect(log.mock.calls[0][0]).toMatchSnapshot();
    });
    
});

describe('Runing command', () => {

    beforeAll(() => {
        jest.clearAllMocks();
        jest.spyOn(Object, 'keys').mockReturnValueOnce([]);
        execa.shellSync.mockReturnValue();
        spyInquirerRobotDefaultValue();
    });

    it('should init project success', async () => {
        await run();
        expect(inquirerRobot.run.mock.calls.length).toEqual(1);
        expect(task.createFiles.mock.calls.length).toEqual(1);
        expect(task.installPackages.mock.calls.length).toEqual(1);
        expect(task.runExtraSettings.mock.calls.length).toEqual(1);
        expect(task.showCompleteMessages.mock.calls.length).toEqual(1);
    });
    
});