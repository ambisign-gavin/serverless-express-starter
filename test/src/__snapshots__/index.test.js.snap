// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runing command with options should show help when unknow options 1`] = `
"
Usage: serverless-express-starter <command>
"
`;

exports[`Runing command with options should show help when unknow options 2`] = `
"where <command> is one of:
"
`;

exports[`Runing command without git should show error when not installed git 1`] = `"serverless-express-started need to install git."`;
